Remove depstar

dpsutton opened the pull request:
tools.build has been updated and depstar has been
[archived](https://github.com/seancorfield/depstar/blob/develop/README.md).

Tools.build copies all of the files into a temp jar whereas depstar would consume
directly from jars. The difference of this raised his head on osx. OSX
is a case insensitive file system whereas jar files are case
insensitive. In our dependency jars we had a situation like

LICENSE   - a file
license/license.txt - a file in a directory called license

In a jar these are fine. On a case sensitive filesystem these can exist
side by side since a file license and a folder LICENSE can
coexist. However on osx this cannot happen since ./license and ./LICENSE
are the same, and one being a file and one a directory is impossible.

Tools.build added exclusions and conflict handlers

```clojure
clojure.tools.build.api/uber
([params])
  Create uberjar file containing contents of deps in basis and class-dir.
  Use main class in manifest if provided. Returns nil.

  Options:
    :class-dir - required, local class dir to include
    :uber-file - required, uber jar file to create
    :basis - used to pull dep jars
    :main - main class symbol
    :manifest - map of manifest attributes, merged last over defaults + :main
    :exclude - coll of string patterns (regex) to exclude from deps
    :conflict-handlers - map of string pattern (regex) to built-in handlers,
                         symbols to eval, or function instances
```

We exclude the single `LICENSE` file so that the other `license/*` files can exist on the filesystem. We still have 3rd party licenses bundled so no issue.

Tools.build also adds ability to pass in manifest options. One thing we have to do though is override the auto-detection of multi-release (#17002) so we end up with `Multi-Release: false` in our jar.

dpsutton committed:
Remove depstar

tools.build has been updated and depstar has been
[archived](https://github.com/seancorfield/depstar/blob/develop/README.md). However,
I'm surprised that this actually worked out of the box. Tools.build
copies all of the files into a temp jar whereas depstar would consume
directly from jars. The difference of this raised his head on osx. OSX
is a case insensitive file system whereas jar files are case
insensitive. In our dependency jars we had a situation like

META-INF/license
META-INF/LICENSE/license.txt

In a jar these are fine. On a case sensitive filesystem these can exist
side by side since a file license and a folder LICENSE can
coexist. However on osx this cannot happen since ./license and ./LICENSE
are the same, and one being a file and one a directory is impossible.

Tools.build added exclusions and conflict handlers

```clojure
clojure.tools.build.api/uber
([params])
  Create uberjar file containing contents of deps in basis and class-dir.
  Use main class in manifest if provided. Returns nil.

  Options:
    :class-dir - required, local class dir to include
    :uber-file - required, uber jar file to create
    :basis - used to pull dep jars
    :main - main class symbol
    :manifest - map of manifest attributes, merged last over defaults + :main
    :exclude - coll of string patterns (regex) to exclude from deps
    :conflict-handlers - map of string pattern (regex) to built-in handlers,
                         symbols to eval, or function instances
```

I was expecting to have to exclude one of the licenses to solve this (it
cannot be fixed by conflict handlers which can append or replace files
with the same name). But there was no conflict. I'm not sure what
changed but it seems to have worked out.

pawit-metabase commented on deps.edn:
```diff
@@ -396,8 +396,7 @@
   :build
-  {:deps       {io.github.clojure/tools.build   {:git/tag "v0.1.6", :git/sha "5636e61"}
-                com.github.seancorfield/depstar {:mvn/version "2.1.278"}
+  {:deps       {io.github.clojure/tools.build   {:git/tag "v0.5.0", :git/sha "7d77952"}
```
this in a few other files too, should we update them all to be on a consistent version?

dpsutton replied:
what is in a few other places? tools.build? If so yes that should for sure be updated. i'll check

pawit-metabase replied:
Yes, tools.build

```
bin/build-drivers/deps.edn
10:  io.github.clojure/tools.build   {:git/tag "v0.1.6", :git/sha "5636e61"}

java/deps.edn
10:  {:deps       {io.github.clojure/tools.build {:git/tag "v0.1.6" :git/sha "5636e61"}}

modules/drivers/sparksql/deps.edn
38:  {:deps       {io.github.clojure/tools.build {:git/tag "v0.1.6" :git/sha "5636e61"}}
```

pawit-metabase reviewed (commented):
Doesn't work on my mac

```
       :cause
       "Cannot write license/LICENSE from xml-apis/xml-apis-ext as parent dir is a file from another lib. One of them must be excluded.",
       :data {}}
          Step "Create uberjar" failed with error "Cannot write license/LICENSE from xml-apis/xml-apis-ext as parent dir is a file from another lib. One of them must be excluded."
```

dpsutton commented:
Yeah i just ran into the failure locally as well. I think i somehow messed up and still had depstar hanging around. I'm not sure how though. But i can recreate now

> Cannot write license/LICENSE from xml-apis/xml-apis-ext as parent dir is a file from another lib. One of them must be excluded.

Reevaluating things in this file is a bit annoying and i'm not sure why. I'll fix and update those other places. Good catch thanks @pawit-metabase

dpsutton committed:
Use tools.build manifest apparatus, use same tools.build everywhere

dpsutton committed:
Exclude fewer files

there is a single file LICENSE and several license/whatever files. If we
exclude the single file we get all of the files in the directory
license/ so that is the better option to keep more and exclude one

tsmacdonald commented:
Closing this since it was fixed elsewhere:

![image](https://user-images.githubusercontent.com/784417/204255766-83f755ef-56e7-4037-88f1-c24af979d9d6.png)

tsmacdonald commented:
(Hmm, should the build-drivers dep be upgraded to 0.7.5?)
